File: ios/Sources/BoardScene/Games.swift
import ChessCore

/// One half-move, with everything the board needs to animate it.
///
/// Nothing here is transcribed. The games are stored as the moves they were
/// actually played in — the notation a book prints — and expanded by the same
/// move generator the rest of the app plays with. That settles en passant,
/// castling and promotion by the rules rather than by hand, and a wrong move
/// in a game list fails to expand instead of quietly animating a piece through
/// a wall.
public struct Ply: Sendable, Equatable {
    public let from: Square
    public let to: Square
    public let kind: PieceKind
    public let color: PieceColor
    /// Where the taken piece was standing — which for en passant is not `to`.
    public let capture: Square?
    public let promotion: PieceKind?
    /// Castling moves a rook as well.
    public let rook: RookMove?
    public let san: String

    public struct RookMove: Sendable, Equatable {
        public let from: Square
        public let to: Square
    }
}

/// A game the board plays out on its own.
public struct ShowGame: Sendable, Identifiable, Equatable {
    public let id: String
    public let white: String
    public let black: String
    public let event: String
    public let year: Int
    public let result: String
    public let note: String
    public let plies: [Ply]

    public var caption: String { "\(white) — \(black)" }
    public var occasion: String { "\(event), \(year) · \(result)" }
}

public enum ShowGames {
    /// Three of the most famous attacking games ever recorded, the same three
    /// the site plays. Each ends in mate and none of them is a grind: a title
    /// sequence has about a minute of anyone's attention and needs every move
    /// to be doing something.
    public static let all: [ShowGame] = [
        make(
            id: "opera", white: "Paul Morphy", black: "Duke of Brunswick & Count Isouard",
            event: "Paris Opera", year: 1858, result: "1–0",
            note: "Seventeen moves, every one of them developing something, and a mate delivered by the last piece to arrive.",
            moves: """
            e4 e5 Nf3 d6 d4 Bg4 dxe5 Bxf3 Qxf3 dxe5 Bc4 Nf6 Qb3 Qe7 Nc3 c6 Bg5 b5 Nxb5 cxb5
            Bxb5+ Nbd7 O-O-O Rd8 Rxd7 Rxd7 Rd1 Qe6 Bxd7+ Nxd7 Qb8+ Nxb8 Rd8#
            """
        ),
        make(
            id: "evergreen", white: "Adolf Anderssen", black: "Jean Dufresne",
            event: "Berlin", year: 1852, result: "1–0",
            note: "A queen given away on move twenty-two, and mate three moves later with the two bishops nobody was watching.",
            moves: """
            e4 e5 Nf3 Nc6 Bc4 Bc5 b4 Bxb4 c3 Ba5 d4 exd4 O-O d3 Qb3 Qf6 e5 Qg6 Re1 Nge7
            Ba3 b5 Qxb5 Rb8 Qa4 Bb6 Nbd2 Bb7 Ne4 Qf5 Bxd3 Qh5 Nf6+ gxf6 exf6 Rg8 Rad1 Qxf3
            Rxe7+ Nxe7 Qxd7+ Kxd7 Bf5+ Ke8 Bd7+ Kf8 Bxe7#
            """
        ),
        make(
            id: "immortal", white: "Adolf Anderssen", black: "Lionel Kieseritzky",
            event: "London", year: 1851, result: "1–0",
            note: "Both rooks and the queen sacrificed, and mate delivered by three minor pieces against a full board.",
            moves: """
            e4 e5 f4 exf4 Bc4 Qh4+ Kf1 b5 Bxb5 Nf6 Nf3 Qh6 d3 Nh5 Nh4 Qg5 Nf5 c6 g4 Nf6
            Rg1 cxb5 h4 Qg6 h5 Qg5 Qf3 Ng8 Bxf4 Qf6 Nc3 Bc5 Nd5 Qxb2 Bd6 Bxg1 e5 Qxa1+
            Ke2 Na6 Nxg7+ Kd8 Qf6+ Nxf6 Be7#
            """
        ),
    ]

    private static func make(
        id: String, white: String, black: String, event: String, year: Int,
        result: String, note: String, moves: String
    ) -> ShowGame {
        ShowGame(
            id: id, white: white, black: black, event: event, year: year,
            result: result, note: note, plies: expand(moves)
        )
    }

    /// Turns a game's notation into plies by playing it.
    ///
    /// A move the generator will not accept stops the expansion there rather
    /// than throwing: a board that plays a famous game two moves short is a
    /// smaller failure than a first screen that crashes, and the tests catch
    /// the short game long before anybody sees it.
    static func expand(_ notation: String) -> [Ply] {
        expandFrom(position: Position(), notation: notation)
    }

    /// The same, from a position that is not the starting one. Used by the
    /// tests to reach the rules these three games never exercise.
    static func expandFrom(position start: Position, notation: String) -> [Ply] {
        var position = start
        var plies: [Ply] = []

        for token in notation.split(whereSeparator: \.isWhitespace) {
            let san = String(token)
            guard let move = position.move(san: san),
                  let piece = position[move.from]
            else { break }

            let capture: Square? = switch move.kind {
            case .enPassant: Square(file: move.to.file, rank: move.from.rank)
            case .capture: move.to
            default: nil
            }

            // The rook's half of a castle, which the notation leaves implicit.
            let rank = move.from.rank
            let rook: Ply.RookMove? = switch move.kind {
            case .kingsideCastle:
                .init(from: Square(file: 7, rank: rank), to: Square(file: 5, rank: rank))
            case .queensideCastle:
                .init(from: Square(file: 0, rank: rank), to: Square(file: 3, rank: rank))
            default: nil
            }

            plies.append(Ply(
                from: move.from, to: move.to, kind: piece.kind, color: piece.color,
                capture: capture, promotion: move.promotion, rook: rook, san: san
            ))

            _ = position.make(move)
        }

        return plies
    }
}

File: web/src/app/three/games.ts
import { PieceKind } from './pieces';

/**
 * The games the title sequence plays out.
 *
 * Three of the most famous attacking games ever recorded, chosen because each
 * one ends in mate and none of them is a grind: a title sequence has about a
 * minute of a viewer's attention and needs every move to be doing something.
 *
 * The move lists are not typed out by hand. They were expanded from PGN by
 * chess.js — every move validated, captures resolved to the square the taken
 * piece actually stood on (which for en passant is not the destination), and
 * castling split into the two pieces that move. If a move in the source PGN
 * had been wrong the expansion would have thrown rather than shipped.
 */

/** One half-move, with everything the board needs to animate it. */
export interface Ply {
  readonly from: string;
  readonly to: string;
  readonly kind: PieceKind;
  readonly white: boolean;
  /** Square the captured piece stands on — not always `to`. */
  readonly capture?: string;
  readonly promotion?: PieceKind;
  /** Castling moves a rook as well. */
  readonly rookFrom?: string;
  readonly rookTo?: string;
  readonly san: string;
}

export interface Game {
  readonly id: string;
  readonly white: string;
  readonly black: string;
  readonly event: string;
  readonly year: number;
  readonly result: string;
  readonly note: string;
  readonly plies: readonly Ply[];
}

export const GAMES: readonly Game[] = [
  {
    id: 'opera',
    white: 'Paul Morphy',
    black: 'Duke of Brunswick & Count Isouard',
    event: 'Paris Opera',
    year: 1858,
    result: '1–0',
    note: 'Seventeen moves, every one of them developing something, and a mate delivered by the last piece to arrive.',
    plies: [
      { from: 'e2', to: 'e4', kind: 'pawn', white: true, san: 'e4' },
      { from: 'e7', to: 'e5', kind: 'pawn', white: false, san: 'e5' },
      { from: 'g1', to: 'f3', kind: 'knight', white: true, san: 'Nf3' },
      { from: 'd7', to: 'd6', kind: 'pawn', white: false, san: 'd6' },
      { from: 'd2', to: 'd4', kind: 'pawn', white: true, san: 'd4' },
      { from: 'c8', to: 'g4', kind: 'bishop', white: false, san: 'Bg4' },
      { from: 'd4', to: 'e5', kind: 'pawn', white: true, capture: 'e5', san: 'dxe5' },
      { from: 'g4', to: 'f3', kind: 'bishop', white: false, capture: 'f3', san: 'Bxf3' },
      { from: 'd1', to: 'f3', kind: 'queen', white: true, capture: 'f3', san: 'Qxf3' },
      { from: 'd6', to: 'e5', kind: 'pawn', white: false, capture: 'e5', san: 'dxe5' },
      { from: 'f1', to: 'c4', kind: 'bishop', white: true, san: 'Bc4' },
      { from: 'g8', to: 'f6', kind: 'knight', white: false, san: 'Nf6' },
      { from: 'f3', to: 'b3', kind: 'queen', white: true, san: 'Qb3' },
      { from: 'd8', to: 'e7', kind: 'queen', white: false, san: 'Qe7' },
      { from: 'b1', to: 'c3', kind: 'knight', white: true, san: 'Nc3' },
      { from: 'c7', to: 'c6', kind: 'pawn', white: false, san: 'c6' },
      { from: 'c1', to: 'g5', kind: 'bishop', white: true, san: 'Bg5' },
      { from: 'b7', to: 'b5', kind: 'pawn', white: false, san: 'b5' },
      { from: 'c3', to: 'b5', kind: 'knight', white: true, capture: 'b5', san: 'Nxb5' },
      { from: 'c6', to: 'b5', kind: 'pawn', white: false, capture: 'b5', san: 'cxb5' },
      { from: 'c4', to: 'b5', kind: 'bishop', white: true, capture: 'b5', san: 'Bxb5+' },
      { from: 'b8', to: 'd7', kind: 'knight', white: false, san: 'Nbd7' },
      {
        from: 'e1',
        to: 'c1',
        kind: 'king',
        white: true,
        rookFrom: 'a1',
        rookTo: 'd1',
        san: 'O-O-O',
      },
      { from: 'a8', to: 'd8', kind: 'rook', white: false, san: 'Rd8' },
      { from: 'd1', to: 'd7', kind: 'rook', white: true, capture: 'd7', san: 'Rxd7' },
      { from: 'd8', to: 'd7', kind: 'rook', white: false, capture: 'd7', san: 'Rxd7' },
      { from: 'h1', to: 'd1', kind: 'rook', white: true, san: 'Rd1' },
      { from: 'e7', to: 'e6', kind: 'queen', white: false, san: 'Qe6' },
      { from: 'b5', to: 'd7', kind: 'bishop', white: true, capture: 'd7', san: 'Bxd7+' },
      { from: 'f6', to: 'd7', kind: 'knight', white: false, capture: 'd7', san: 'Nxd7' },
      { from: 'b3', to: 'b8', kind: 'queen', white: true, san: 'Qb8+' },
      { from: 'd7', to: 'b8', kind: 'knight', white: false, capture: 'b8', san: 'Nxb8' },
      { from: 'd1', to: 'd8', kind: 'rook', white: true, san: 'Rd8#' },
    ],
  },
  {
    id: 'evergreen',
    white: 'Adolf Anderssen',
    black: 'Jean Dufresne',
    event: 'Berlin',
    year: 1852,
    result: '1–0',
    note: 'The Evergreen. A queen given away on move nineteen for a mate that takes five more.',
    plies: [
      { from: 'e2', to: 'e4', kind: 'pawn', white: true, san: 'e4' },
      { from: 'e7', to: 'e5', kind: 'pawn', white: false, san: 'e5' },
      { from: 'g1', to: 'f3', kind: 'knight', white: true, san: 'Nf3' },
      { from: 'b8', to: 'c6', kind: 'knight', white: false, san: 'Nc6' },
      { from: 'f1', to: 'c4', kind: 'bishop', white: true, san: 'Bc4' },
      { from: 'f8', to: 'c5', kind: 'bishop', white: false, san: 'Bc5' },
      { from: 'b2', to: 'b4', kind: 'pawn', white: true, san: 'b4' },
      { from: 'c5', to: 'b4', kind: 'bishop', white: false, capture: 'b4', san: 'Bxb4' },
      { from: 'c2', to: 'c3', kind: 'pawn', white: true, san: 'c3' },
      { from: 'b4', to: 'a5', kind: 'bishop', white: false, san: 'Ba5' },
      { from: 'd2', to: 'd4', kind: 'pawn', white: true, san: 'd4' },
      { from: 'e5', to: 'd4', kind: 'pawn', white: false, capture: 'd4', san: 'exd4' },
      { from: 'e1', to: 'g1', kind: 'king', white: true, rookFrom: 'h1', rookTo: 'f1', san: 'O-O' },
      { from: 'd4', to: 'd3', kind: 'pawn', white: false, san: 'd3' },
      { from: 'd1', to: 'b3', kind: 'queen', white: true, san: 'Qb3' },
      { from: 'd8', to: 'f6', kind: 'queen', white: false, san: 'Qf6' },
      { from: 'e4', to: 'e5', kind: 'pawn', white: true, san: 'e5' },
      { from: 'f6', to: 'g6', kind: 'queen', white: false, san: 'Qg6' },
      { from: 'f1', to: 'e1', kind: 'rook', white: true, san: 'Re1' },
      { from: 'g8', to: 'e7', kind: 'knight', white: false, san: 'Nge7' },
      { from: 'c1', to: 'a3', kind: 'bishop', white: true, san: 'Ba3' },
      { from: 'b7', to: 'b5', kind: 'pawn', white: false, san: 'b5' },
      { from: 'b3', to: 'b5', kind: 'queen', white: true, capture: 'b5', san: 'Qxb5' },
      { from: 'a8', to: 'b8', kind: 'rook', white: false, san: 'Rb8' },
      { from: 'b5', to: 'a4', kind: 'queen', white: true, san: 'Qa4' },
      { from: 'a5', to: 'b6', kind: 'bishop', white: false, san: 'Bb6' },
      { from: 'b1', to: 'd2', kind: 'knight', white: true, san: 'Nbd2' },
      { from: 'c8', to: 'b7', kind: 'bishop', white: false, san: 'Bb7' },
      { from: 'd2', to: 'e4', kind: 'knight', white: true, san: 'Ne4' },
      { from: 'g6', to: 'f5', kind: 'queen', white: false, san: 'Qf5' },
      { from: 'c4', to: 'd3', kind: 'bishop', white: true, capture: 'd3', san: 'Bxd3' },
      { from: 'f5', to: 'h5', kind: 'queen', white: false, san: 'Qh5' },
      { from: 'e4', to: 'f6', kind: 'knight', white: true, san: 'Nf6+' },
      { from: 'g7', to: 'f6', kind: 'pawn', white: false, capture: 'f6', san: 'gxf6' },
      { from: 'e5', to: 'f6', kind: 'pawn', white: true, capture: 'f6', san: 'exf6' },
      { from: 'h8', to: 'g8', kind: 'rook', white: false, san: 'Rg8' },
      { from: 'a1', to: 'd1', kind: 'rook', white: true, san: 'Rad1' },
      { from: 'h5', to: 'f3', kind: 'queen', white: false, capture: 'f3', san: 'Qxf3' },
      { from: 'e1', to: 'e7', kind: 'rook', white: true, capture: 'e7', san: 'Rxe7+' },
      { from: 'c6', to: 'e7', kind: 'knight', white: false, capture: 'e7', san: 'Nxe7' },
      { from: 'a4', to: 'd7', kind: 'queen', white: true, capture: 'd7', san: 'Qxd7+' },
      { from: 'e8', to: 'd7', kind: 'king', white: false, capture: 'd7', san: 'Kxd7' },
      { from: 'd3', to: 'f5', kind: 'bishop', white: true, san: 'Bf5+' },
      { from: 'd7', to: 'e8', kind: 'king', white: false, san: 'Ke8' },
      { from: 'f5', to: 'd7', kind: 'bishop', white: true, san: 'Bd7+' },
      { from: 'e8', to: 'f8', kind: 'king', white: false, san: 'Kf8' },
      { from: 'a3', to: 'e7', kind: 'bishop', white: true, capture: 'e7', san: 'Bxe7#' },
    ],
  },
  {
    id: 'immortal',
    white: 'Adolf Anderssen',
    black: 'Lionel Kieseritzky',
    event: 'London',
    year: 1851,
    result: '1–0',
    note: 'The Immortal. Both rooks and the queen given away, and mate with the three pieces left.',
    plies: [
      { from: 'e2', to: 'e4', kind: 'pawn', white: true, san: 'e4' },
      { from: 'e7', to: 'e5', kind: 'pawn', white: false, san: 'e5' },
      { from: 'f2', to: 'f4', kind: 'pawn', white: true, san: 'f4' },
      { from: 'e5', to: 'f4', kind: 'pawn', white: false, capture: 'f4', san: 'exf4' },
      { from: 'f1', to: 'c4', kind: 'bishop', white: true, san: 'Bc4' },
      { from: 'd8', to: 'h4', kind: 'queen', white: false, san: 'Qh4+' },
      { from: 'e1', to: 'f1', kind: 'king', white: true, san: 'Kf1' },
      { from: 'b7', to: 'b5', kind: 'pawn', white: false, san: 'b5' },
      { from: 'c4', to: 'b5', kind: 'bishop', white: true, capture: 'b5', san: 'Bxb5' },
      { from: 'g8', to: 'f6', kind: 'knight', white: false, san: 'Nf6' },
      { from: 'g1', to: 'f3', kind: 'knight', white: true, san: 'Nf3' },
      { from: 'h4', to: 'h6', kind: 'queen', white: false, san: 'Qh6' },
      { from: 'd2', to: 'd3', kind: 'pawn', white: true, san: 'd3' },
      { from: 'f6', to: 'h5', kind: 'knight', white: false, san: 'Nh5' },
      { from: 'f3', to: 'h4', kind: 'knight', white: true, san: 'Nh4' },
      { from: 'h6', to: 'g5', kind: 'queen', white: false, san: 'Qg5' },
      { from: 'h4', to: 'f5', kind: 'knight', white: true, san: 'Nf5' },
      { from: 'c7', to: 'c6', kind: 'pawn', white: false, san: 'c6' },
      { from: 'g2', to: 'g4', kind: 'pawn', white: true, san: 'g4' },
      { from: 'h5', to: 'f6', kind: 'knight', white: false, san: 'Nf6' },
      { from: 'h1', to: 'g1', kind: 'rook', white: true, san: 'Rg1' },
      { from: 'c6', to: 'b5', kind: 'pawn', white: false, capture: 'b5', san: 'cxb5' },
      { from: 'h2', to: 'h4', kind: 'pawn', white: true, san: 'h4' },
      { from: 'g5', to: 'g6', kind: 'queen', white: false, san: 'Qg6' },
      { from: 'h4', to: 'h5', kind: 'pawn', white: true, san: 'h5' },
      { from: 'g6', to: 'g5', kind: 'queen', white: false, san: 'Qg5' },
      { from: 'd1', to: 'f3', kind: 'queen', white: true, san: 'Qf3' },
      { from: 'f6', to: 'g8', kind: 'knight', white: false, san: 'Ng8' },
      { from: 'c1', to: 'f4', kind: 'bishop', white: true, capture: 'f4', san: 'Bxf4' },
      { from: 'g5', to: 'f6', kind: 'queen', white: false, san: 'Qf6' },
      { from: 'b1', to: 'c3', kind: 'knight', white: true, san: 'Nc3' },
      { from: 'f8', to: 'c5', kind: 'bishop', white: false, san: 'Bc5' },
      { from: 'c3', to: 'd5', kind: 'knight', white: true, san: 'Nd5' },
      { from: 'f6', to: 'b2', kind: 'queen', white: false, capture: 'b2', san: 'Qxb2' },
      { from: 'f4', to: 'd6', kind: 'bishop', white: true, san: 'Bd6' },
      { from: 'c5', to: 'g1', kind: 'bishop', white: false, capture: 'g1', san: 'Bxg1' },
      { from: 'e4', to: 'e5', kind: 'pawn', white: true, san: 'e5' },
      { from: 'b2', to: 'a1', kind: 'queen', white: false, capture: 'a1', san: 'Qxa1+' },
      { from: 'f1', to: 'e2', kind: 'king', white: true, san: 'Ke2' },
      { from: 'b8', to: 'a6', kind: 'knight', white: false, san: 'Na6' },
      { from: 'f5', to: 'g7', kind: 'knight', white: true, capture: 'g7', san: 'Nxg7+' },
      { from: 'e8', to: 'd8', kind: 'king', white: false, san: 'Kd8' },
      { from: 'f3', to: 'f6', kind: 'queen', white: true, san: 'Qf6+' },
      { from: 'g8', to: 'f6', kind: 'knight', white: false, capture: 'f6', san: 'Nxf6' },
      { from: 'd6', to: 'e7', kind: 'bishop', white: true, san: 'Be7#' },
    ],
  },
];
